feat: add `unstable_htmlAsRichText` and `unstable_markdownAsRichText` helpers

lihbr opened the pull request:
<!-- Please use a Conventional Commit in your PR title -->
<!-- https://conventionalcommits.org -->
<!-- e.g. "feat: support new field type" -->

**Implements**: [RFC: HTML to Rich Text Serializer](https://www.notion.so/prismic/RFC-HTML-to-Rich-Text-Serializer-03208612db93475e8409f528a0194b91?pvs=9)
**Fixes**: #331

### Description

This pull request implements helpers to serialize HTML and markdown to Prismic rich text field format as described per the related RFC.

These new helpers are introduced to help users migrating to Prismic through the migration API to convert their HTML content to rich text easily.

### Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- Don't hesitate to ask for help! -->

- [x] If my changes require tests, I added them.
- [x] If my changes affect backward compatibility, it has been discussed.
- [x] If my changes require an update to the CONTRIBUTING.md guide, I updated it.
- [x] Discuss what we want to expose from where
   - [x] Where to export `htmlAsRichText`, `markdownAsRichText`, and `AsRichTextConfig` from
   - [x] Whether or not to export `filterRichTextField` (could be helpful internally)
   - [x] Whether or not to export `rehypeRichText` and `hastUtilToRichText`

### How to QA [^1]

<!-- When relevant, describe how to QA your changes. -->
- Review the test suite introduced by this pull request: https://github.com/prismicio/prismic-client/blob/lh/html-to-richtext-converter/test/richtext/htmlAsRichText.test.ts
- Try it yourself creating a playground file on a clone of the repo and running it with sample code
   ```typescript
   // run with `npx tsx playground.ts`
   import { unstable_htmlAsRichText } from "./src";

   const result = htmlAsRichText(/* html */ `<p>hello world</p>`);

   console.log(result);
   ```

<!-- Your favorite emoji is welcome to close your PR! -->

<!-- A note for reviewers: -->

[^1]:
	Please use these labels when submitting a review:
	:question: #ask:&ensp;Ask a question.
	:bulb: #idea:&ensp;Suggest an idea.
	:warning: #issue:&ensp;Strongly suggest a change.
	:tada: #nice:&ensp;Share a compliment.

angeloashmore reviewed (changes requested):
Awesome work! It seems to work really well — not an easy feat.

I'm going to mark the PR as "Request changes" since there are a number of comments that require discussion, but that doesn't necessarily mean anything needs to change. I don't have any blocking comments.

## ❓ General questions

- Returning `messages` as `VFileMessage[]` ties our API to unified. Although we don't have any immediate plans to transition away from unified, I could see us wanting to explore a version without it for performance and size benefits. What do you think of abstracting `VFileMessage[]` into a higher-level version with objects that we create?

- Should `warnings` be named `messages` to match unified's output? I wonder if there would ever be a message that isn't a warning, in which case `warnings` would not make sense. If the messages are always warnings, `warnings` is good.

- `filterRichTextField` seems like something that belongs in user code rather than `@prismicio/client`. It functions as a way to clean up already existing rich text data. It assumes the user has easy access to the model and is working with invalid rich text data for the model. Couldn't the user simply not return invalid rich text nodes in the serializer?

  - What happens when rich text data that contains invalid nodes (e.g. a `heading1` when Heading 1 is not enabled) is uploaded to Prismic? Does it filter out the invalid nodes automatically? If yes, `filterRichTextField` may not be necessary for the `*AsRichText` helper's intended use cases.

lihbr commented on src/richtext/htmlAsRichText.ts:
```diff
@@ -0,0 +1,29 @@
+ *
+ * @returns Rich text field equivalent of the provided HTML string.
+ *
+ * @experimental - This API is subject to change and might not follow SemVer.
```
I think my main worry is that we could realize we need those functions to be async after all, otherwise it's supposed to be stable to me.

angeloashmore replied:
I see, that makes sense. Since we plan to release this helper before our work on the Migration API client, it might be better to explicitly mark the helper as unstable.

We could export the helper as `unstable_htmlAsRichText`/`unstable_markdownAsRichText`. Then, when we are more confident in the helper and its compatibility with the Migration API, we can remove `unstable_`.

I worry users won't see the `@experimental` tag in the TSDocs and expect the function to follow SemVer.

lihbr replied:
Happy to go that way!

lihbr replied:
Updated to `unstable_`

lihbr commented on src/richtext/types.ts:
```diff
@@ -1,3 +1,5 @@
+import { VFile } from "vfile";
```
:100: I was actually surprised we didn't do it already.

lihbr replied:
I'll enable `"@typescript-eslint/consistent-type-imports": "error",` before merging to reduce the noise before (100+ files changed)

angeloashmore replied:
How about we do it in a separate PR after this is merged so this one stays clean? It's mainly an aesthetic improvement so not necessary for this PR.

lihbr commented on src/richtext/utils/RichTextFieldBuilder.ts:
```diff
@@ -0,0 +1,159 @@
+ *
+ * @see https://github.com/sindresorhus/type-fest/blob/8a45ba048767aaffcebc7d190172d814a739feb0/source/distributed-omit.d.ts
+ */
+type DistributedOmit<
```
I tried for `DistributedOmit` and `DistributedPick` but couldn't replicate their TypeScript black magic correctly unfortunately. But yeah I don't like them.

angeloashmore replied:
That `extends unknown` is carrying a lot of weight. 😅

This slightly simpler version seems to work the same for our use case:

```ts
type DistributedOmit<
	ObjectType,
	KeyType extends keyof ObjectType,
> = ObjectType extends unknown ? Omit<ObjectType, KeyType> : never;
```

I'll let you decide if you want to use that version of the original one. The original probably handles some edge cases.

lihbr replied:
Made the small refactor, thanks for looking into it!

lihbr commented on src/richtext/utils/hastUtilToRichText.ts:
```diff
@@ -0,0 +1,373 @@
+	| RTNode
+	| RTInlineNode
+	| RTPartialInlineNode
+	| null
```
That was to be consistent with the rich text map serializer equivalent, but we could get rid of it.

lihbr replied:
Removed!

lihbr commented on src/richtext/utils/hastUtilToRichText.ts:
```diff
@@ -0,0 +1,373 @@
+	em: "em",
+	li: ({ context }) =>
+		context.listType === "group-o-list-item" ? "o-list-item" : "list-item",
+	ul: "group-list-item",
+	ol: "group-o-list-item",
```
They can only return the shorthand equivalent (I'll double check, but you cannot return a list node from a function serializer). This is just used to keep track of the last list type in the tree we encountered.

angeloashmore replied:
Perfect 👌

lihbr commented on src/richtext/utils/hastUtilToRichText.ts:
```diff
@@ -0,0 +1,373 @@
+	defaultWrapperNodeType?: RTTextNode["type"];
+};
+
+export const hastUtilToRichText = (
```
Happy to rename it, this was an attempt at being somewhat consistent with what hast is doing for `hast-util-to-html`: https://github.com/syntax-tree/hast-util-to-html

lihbr replied:
Renamed the export to `toRichText` to align with `hast-util-to-html` while keeping the same file name, let me know what you think!

angeloashmore replied:
`hastToRichText` is clearer in my opinion, along with a `hastToRichText.ts` filename. I think `toHtml` works for `hast-util-to-html` because the function name has the `hast-util-to-html` package name as context. In our case, we don't have the same kind of limited context.

lihbr commented on src/richtext/utils/hastUtilToRichText.ts:
```diff
@@ -0,0 +1,373 @@
+	// Keep track of the last text node type to append text nodes to in
+	// case of an image or an embed node is present inside a paragraph.
+	let lastRTTextNodeType: RTTextNode["type"] =
+		config?.defaultWrapperNodeType ?? RichTextNodeType.paragraph;
```
I believe they can since images are inline tags, but since we can't ensure a `heading1` node to be necessarily represented by an `h1` tag (it could be a `div` or anything), technically this makes it valid HTML, so this is all an escape hatch for such cases.

lihbr replied:
I believe they can since images are inline tags, but since we can't ensure a `heading1` node to be necessarily represented by an `h1` tag (it could be a `div` or anything), technically this makes it valid HTML, so this is all an escape hatch for such cases.

angeloashmore replied:
I'm not clear on the use case of `defaultWrapperNodeType`. I see how it affects the output and that it's really only for cases where you don't have a preceding text node, but I'm not sure where a user would want something besides a paragraph.

Do you have an idea?

angeloashmore replied:
I'm not strongly against the option, but I'm wondering if defaulting to `paragraph` is sufficient.

lihbr replied:
I introduced it after seeing some WordPress ACF fields could only return the inner HTML content of a node. (Example was for an author bio here on the get method example: https://www.advancedcustomfields.com/resources/wp-rest-api-integration/)

I agree however I don't see what else of a paragraph you'd want here, so maybe we can remove it and default to a paragraph node, and later reintroduce it if it's actually needed.

lihbr replied:
Removed for now!

lihbr commented on src/richtext/utils/hastUtilToRichText.ts:
```diff
@@ -0,0 +1,373 @@
+				| RichTextHTMLMapSerializerShorthand
+				| RichTextHTMLMapSerializerFunction;
+
+			// We give priority to CSS selectors over tag names.
+			if (node.matchesSerializer && node.matchesSerializer in serializer) {
+				serializerOrShorthand = serializer[node.matchesSerializer];
+			} else if (node.tagName in serializer) {
+				serializerOrShorthand = serializer[node.tagName];
+			} else {
+				return;
+			}
```
Yes, basically I only populate nodes with `matchSerializer` when it's a CSS selector other than a tag name, because matching those is costly (we need to walk the whole tree once for **each** non-tag name CSS selectors). Matching tag names is cheap and should be preferred when possible like here (note the default serializer only uses tag names)

I'll see if I there's an opportunity when walking the tree previously to populate it by default with `matchSerializer` based on the tag name, otherwise this might be the best place to do it.

angeloashmore replied:
Sounds good, that what I was expecting. Let's leave it as is. 🙂

lihbr commented on src/richtext/types.ts:
```diff
@@ -151,3 +157,16 @@ export const RichTextReversedNodeType = {
+/**
+ * Configuration that determines the output of `*AsRichText` functions.
+ */
+export type AsRichTextConfig = RehypeRichTextConfig;
```
Agree! I separated them ☺️

lihbr commented on src/richtext/filterRichTextField.ts:
The thing is that the serializer cannot be "partial", it has to make sense of most node types to correctly process the tree as its being walked through, and then we can filter the rich text output.

For example, if the following serializer: `{ h1: "heading1" }` wasn't merged against the default serializer (which defines defaults for other common tags), processing this input:

```html
<h1>hello world</h1>
<p>lorem ipsum dolor sit amet</p>
```

Would yield this rich text output:

```json
[
	{
		"type": "heading1",
		"text": "hello world lorem ipsum dolor sit amet",
		"spans": []
	}
]
```

This is a limitation of walking the tree where you don't know if the node you're currently processing is a child of the previous one or not. I thought about working with some kind of "depth" mechanism, perhaps I should explore that again, but I'm unsure it'll be satisfactory.

angeloashmore replied:
What happens with a serializer like `{ h1: () => undefined }`? I would assume it would omit the `h1` element and its contents, stopping the walker from going deeper.

If that's how it works, a user could define a serializer that omits unsupported RT node types.

(I'm not necessarily suggesting that's a good DX. Just trying to get a sense of what's possible.)

lihbr replied:
I think we could support it that way indeed (as there's a way to skip visiting part of a tree), but wouldn't the `exclude` option be more appropriate then?

I'm not sure we want to get rid of the default serializer, also, having people consciously overriding it for that purpose while it being hidden (as in, you have to look through the source code or doc to know what it handles) feels a bit weird :thinking:

angeloashmore replied:
We decided to keep the serializer logic as it exists today. Consumers that want to exclude specific nodes can use the `exclude` option, which accepts complex CSS selectors.

We can restore conditional exclusion via functions in the future should there be a need.

lihbr commented on src/richtext/htmlAsRichText.ts:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Converts an HTML string to a rich text field.
+ *
+ * @param html - An HTML string
```
Updated most of them to end with periods. The codebase is pretty inconsistent in that regard but we'll fix it along the way 👍

angeloashmore replied:
Ha, I thought we were better about it but yes, we'll fix other cases as we touch them. 🙂

lihbr commented on src/richtext/htmlAsRichText.ts:
```diff
@@ -0,0 +1,29 @@
+ * @param config - Configuration that determines the output of
+ *   `htmlAsRichText()`
+ *
+ * @returns Rich text field equivalent of the provided HTML string.
```
Done

lihbr commented on src/richtext/index.ts:
```diff
@@ -4,11 +4,17 @@ export { asText } from "./asText";
 	RichTextMapSerializerFunction,
 } from "./types";
+
+export type { RichTextHTMLMapSerializer } from "./utils/hastUtilToRichText";
```
Updated!

lihbr commented on src/richtext/types.ts:
```diff
@@ -151,3 +157,16 @@ export const RichTextReversedNodeType = {
+/**
+ * The return type of `*AsRichText` functions.
+ */
+export type AsRichTextReturnType = {
```
Done

lihbr commented on src/richtext/utils/RichTextFieldBuilder.ts:
```diff
@@ -0,0 +1,159 @@
+>;
+
+export class RichTextFieldBuilder {
+	private nodes: RTNode[] = [];
```
Refactored to `#nodes` and such, no strong opinion on it~

lihbr commented on src/richtext/utils/RichTextFieldBuilder.ts:
```diff
@@ -0,0 +1,159 @@
+export class RichTextFieldBuilder {
+	private nodes: RTNode[] = [];
+
+	private get tail(): RTNode | undefined {
```
Refactored! My English failed me~

lihbr commented on src/richtext/utils/RichTextFieldBuilder.ts:
```diff
@@ -0,0 +1,159 @@
+
+		this.nodes.push(node);
+
+		return this;
```
Refactored, I always go fancy when I start implementing something only then to realize it's not needed~

angeloashmore replied:
I do the same, but @bapmrl taught me the value of minimizing returns and exports. 🙂

lihbr commented on src/richtext/utils/RichTextFieldBuilder.ts:
```diff
@@ -0,0 +1,159 @@
+		return this;
+	}
+
+	appendTextNode(type: RTTextNode["type"], direction?: "ltr" | "rtl"): this {
+		return this.appendNode({
+			type: type,
+			text: "",
+			spans: [],
+			direction: direction || "ltr",
+		});
+	}
```
Nice catch, thanks!

lihbr commented on src/richtext/utils/RichTextFieldBuilder.ts:
```diff
@@ -0,0 +1,159 @@
+		// Ensure that the last text node is trimmed.
+		this.cleanupTail();
+
+		// Because `RichTextField` is defined as a tuple,
```
Thanks!

lihbr commented on src/richtext/utils/RichTextFieldBuilder.ts:
```diff
@@ -0,0 +1,159 @@
+
+	appendText(text: string): this {
+		if (!this.isOfKindText(this.tail)) {
+			throw new Error("Cannot append text to non-text tail node");
```
I'm happy to do it, but `PrismicError` requires a URL and a response object. Do we want to make both optional now? Or do you have something else in mind? 🤔

angeloashmore replied:
Ah, good point. I forgot `PrismicError` is designed as a fallback error for `fetch` calls. It would be better to make `PrismicError` a simple extension of `Error` and create a new `PrismicNetworkError` or something, but that would be a breaking change.

Perhaps we could use a new `PrismicRichTextError` in this case. We can refactor that error to extend `PrismicError` in the future when it is simplified.

lihbr replied:
Refactored!

lihbr commented on src/richtext/utils/RichTextFieldBuilder.ts:
```diff
@@ -0,0 +1,159 @@
+			throw new Error("Cannot append text to non-text tail node");
+		}
+
+		if (!this.tail.text) {
+			this.tail.text = text.trimStart();
+		} else {
+			this.tail.text += text;
+		}
```
Makes sense! (was thinking chronologically here)

lihbr commented on src/richtext/utils/hastUtilToRichText.ts:
```diff
@@ -0,0 +1,373 @@
+ *
+ * @see https://github.com/sindresorhus/type-fest/blob/8a45ba048767aaffcebc7d190172d814a739feb0/source/distributed-pick.d.ts
+ */
+type DistributedPick<
```
Answered in the other thread, we'll continue there!

lihbr commented on src/richtext/utils/hastUtilToRichText.ts:
```diff
@@ -0,0 +1,373 @@
+	| null
+	| undefined;
+
+/**
+ * Serializes an hast {@link Element} node matching the given HTML tag name or
+ * CSS selector to a Prismic rich text node.
+ */
```
Searched and replaced them all!

lihbr commented on src/richtext/utils/hastUtilToRichText.ts:
```diff
@@ -0,0 +1,373 @@
+
+	// Keep track of the last list type to know whether we need to append
+	// `list-item` or `o-list-item` nodes.
+	let lastListType: "group-list-item" | "group-o-list-item" | null = null;
```
Refactored to use `undefined`

lihbr commented on src/richtext/utils/rehypeRichText.ts:
```diff
@@ -0,0 +1,189 @@
+} & HastUtilToRichTextConfig;
+
+/**
+ * A unified plugin that compiles an hast tree to a Prismic rich text field.
```
Updated!

lihbr commented on src/richtext/utils/rehypeRichText.ts:
```diff
@@ -0,0 +1,189 @@
+			// Mark nodes matching CSS selectors
+			if (config?.serializer) {
+				for (const key in config.serializer) {
+					// HTML tag names are single word, lowercase strings, a-z and 1-6 (headings).
+					// Here we want to match anything that's not a valid HTML tag name and treat
+					// it as a CSS selector. See: https://regex101.com/r/LILLWH/1
```
Answered in the above, will continue discussion there!

lihbr commented on src/richtext/utils/serializerHelpers.ts:
Renamed to `hastserializerHelpers`

lihbr commented on src/richtext/utils/serializerHelpers.ts:
```diff
@@ -0,0 +1,158 @@
+ * A warning thrown when a serializer encounters an issue while working with an
+ * hast {@link Element} node. It's then added to the vfile messages.
+ */
+export class SerializerWarning extends Error {}
```
Happy to do it but same issue as above: https://github.com/prismicio/prismic-client/pull/342#discussion_r1684457360

angeloashmore replied:
Following up on my reply there, maybe we could extend from a new `PrismicRichTextError`, which itself could extend `PrismicError` when we make it more generic in the future.

Maybe `PrismicRichTextSerializerError` makes more sense than `SerializerError` in that context.

lihbr replied:
Refactored!

lihbr commented on src/richtext/utils/serializerHelpers.ts:
```diff
@@ -0,0 +1,158 @@
+
+	const url = new URL(src);
+
+	let width = (node.properties?.width as number) ?? 0;
+	let height = (node.properties?.height as number) ?? 0;
```
I updated them to `undefined` but I had to cast them as `number` down the road because the API types are always filled. Does that sound OK to you? Should we update the API types instead to allow `undefined`?

angeloashmore replied:
Hm… the public API always returns dimension values so we should keep them as always defined.

How do we know the Migration API doesn't require them? If the API really doesn't require them, then casting is good with me. When we start working on the Migration API client, we may want to look into a custom Migration API-specific `ImageField` type.

lihbr replied:
Updated to cast! That's what @Heyrows shared with me for reference: https://github.com/prismicio/migration-api/blob/5a8455118c5fcbcb60225a94919a97cc7646fd73/src/models/import/validators/fields/nestable/Image/merge.ts#L35-L38, they leave them empty and compute them later on.

I don't have an issue with having a separate `ImageField` type, but then that means we also need to make the rich text types and nodes more flexible to accommodate both or duplicate them, which I feel meh about 🤔

lihbr replied:
Added a small comment about it linking to this thread 👌

lihbr commented on test/__testutils__/testAsRichTextHelper.ts:
```diff
@@ -0,0 +1,91 @@
+	 * output to exactly match the input. This flag can be used to tell it to
+	 * expect the output to not match the input instead.
+	 */
+	expectAsHTMLNotToMatchInput?: boolean;
```
Done!

lihbr commented on src/richtext/utils/RichTextFieldBuilder.ts:
```diff
@@ -0,0 +1,159 @@
+	 *
+	 * @returns `true` if `node` is of kind text, `false` otherwise.
+	 */
+	private isOfKindText(node?: RTNode): node is RTTextNode {
```
Updated!

lihbr commented on src/richtext/utils/RichTextFieldBuilder.ts:
```diff
@@ -0,0 +1,159 @@
+		return this;
+	}
+
+	appendSpanOfLength(partialSpan: RTPartialInlineNode, length: number): this {
```
I refactored it into `appendSpan` with overloads, let me know if you prefer it.

angeloashmore replied:
Looks good! 👍

lihbr commented:
Thank you so much for the prompt review! I updated everything as per it and answered discussions ☺️  

> * Returning `messages` as `VFileMessage[]` ties our API to unified. Although we don't have any immediate plans to transition away from unified, I could see us wanting to explore a version without it for performance and size benefits. What do you think of abstracting `VFileMessage[]` into a higher-level version with objects that we create?

I updated the signature to be an array of string :)

> * Should `warnings` be named `messages` to match unified's output? I wonder if there would ever be a message that isn't a warning, in which case `warnings` would not make sense. If the messages are always warnings, `warnings` is good.

`VFile` has 3 levels of log: info (regular log), messages (warnings), and fatal (error). I think warnings makes it clearer in our case.

> * `filterRichTextField` seems like something that belongs in user code rather than `@prismicio/client`. It functions as a way to clean up already existing rich text data. It assumes the user has easy access to the model and is working with invalid rich text data for the model. Couldn't the user simply not return invalid rich text nodes in the serializer?

Discussed here: https://github.com/prismicio/prismic-client/pull/342#discussion_r1684196698, this is a bit trickier than not providing a serializer for the given type of node I'm afraid.

>   * What happens when rich text data that contains invalid nodes (e.g. a `heading1` when Heading 1 is not enabled) is uploaded to Prismic? Does it filter out the invalid nodes automatically? If yes, `filterRichTextField` may not be necessary for the `*AsRichText` helper's intended use cases.

We'll discover that pretty soon as I start to work with the migration API I assume 🤔

angeloashmore commented on src/richtext/utils/hastUtilToRichText.ts:
```diff
@@ -0,0 +1,373 @@
+	RichTextHTMLMapSerializerShorthand | RichTextHTMLMapSerializerFunction
+>;
+
+const DEFAULT_SERIALIZER: RichTextHTMLMapSerializer = {
```
👏

lihbr commented:
We agreed to export everything from `@prismicio/client` directly.
- `richtext/utils` will be moved to `lib`
- `filterRichTextField.ts` will be moved to `lib`

angeloashmore commented on src/richtext/utils/RichTextFieldBuilder.ts:
```diff
@@ -91,69 +112,53 @@ export class RichTextFieldBuilder {
-		// Because `RichTextField` is defined as a tuple,
-		// we have to cast `RTNode[]` to `RichTextField`.
-		return this.nodes as RichTextField;
+		// Because `RichTextField` is defined as a non-empty,
```
Small fix:

```suggestion
		// Because `RichTextField` is defined as a non-empty
```

angeloashmore commented on src/richtext/utils/RichTextFieldBuilder.ts:
```diff
@@ -91,69 +112,53 @@ export class RichTextFieldBuilder {
-		// Because `RichTextField` is defined as a tuple,
-		// we have to cast `RTNode[]` to `RichTextField`.
-		return this.nodes as RichTextField;
+		// Because `RichTextField` is defined as a non-empty,
+		// array we have to cast `RTNode[]` to `RichTextField`.
```
Small fix:

```suggestion
		// Because `RichTextField` is defined as a non-empty
		// array, we have to cast `RTNode[]` to `RichTextField`.
```

angeloashmore commented on src/richtext/utils/hastToRichText.ts:
```diff
@@ -127,9 +63,9 @@ const VFILE_RULE = "failed-to-serialize-node";
+ * Configuration that determines the output of `toRichText`.
  */
-export type HastUtilToRichTextConfig = {
+export type HASTToRichTextConfig = {
```
💡 #idea: I think this type should be written as `HastToRightTextConfig` since hast is not an acronym like HTML.

lihbr replied:
From GitHub, it's the acronym for Hypertext Abstract Syntax Tree, so I'll leave it as-is, but we can rename it later.

See repository description: https://github.com/syntax-tree/hast

angeloashmore reviewed (approved):
This looks great! ❤️

As we discussed, we just need to export the helpers from the right entries and reorganize some of the `lib` functions.

Tree-shaking works as we want: there are no traces of the unified ecosystem if you don't use the `*AsRichText` helpers. ~48 KB was added when using the `markdownAsRichText` helper. As a general rule, we'll recommend against using the helper on the client.
